packages/server/src/services/rate.service.js

```diff
@@ -0,0 +1,383 @@
+    }
+
+    if (!timestamp) {
+      throw new HttpError(400, "Timestamp is missing");
+    }
+
+    if (!version) {
+      throw new HttpError(400, "Version is missing");
+    }
+
+    const rateCol = await getRateCollection();
+
+    const existRate = await rateCol.findOne({ indexer, address });
+    if (existRate) {
+      throw new HttpError(400, "You had already rated");
+    }
+
+    const pinResult = await pinJsonToIpfs({ msg, address, signature, version: "1" });
+    if (!pinResult.PinHash) {
+      throw new HttpError(500, "Failed to pin to ipfs");
+    }
+
+    const now = new Date();
+    const result = await rateCol.insertOne(
+      {
+        indexer,
+        data,
+        address,
+        signature,
+        pinHash: pinResult.PinHash,
+        version: "1",
+        createdAt: now,
+        updatedAt: now,
+      });
+
+    if (!result.result.ok) {
+      throw new HttpError(500, "Add rate error");
+    }
+
+    const rateId = result.ops[0]._id;
+
+    return rateId;
+  }
+
+  async getRates(indexer, page, pageSize, user) {
+    const rateCol = await getRateCollection();
+
+    const q = {
+      indexer,
+    };
+
+    const total = await rateCol.countDocuments(q);
+
+    if (page === "last") {
+      const totalPages = Math.ceil(total / pageSize);
+      page = totalPages - 1;
+    }
+
+    const rates = await rateCol.aggregate([
+      { $match: q },
+      { $sort: { createdAt: 1 } },
+      { $skip: page * pageSize },
+      { $limit: pageSize },
+      {
+        $lookup: {
+          from: "reaction",
+          let: { rateId: "$_id" },
+          pipeline: [
+            {
+              $match: {
+                $expr: {
+                  $eq: ["$rateId", "$$rateId"],
+                },
+              },
+            },
+            {
+              $group: {
+                _id: "$reaction",
+                count: { $sum: 1 },
+              },
+            },
+            {
+              $project: {
+                _id: 0,
+                reaction: "$_id",
+                count: 1,
+              },
+            },
+          ],
+          as: "reactions",
+        },
+      },
+      ...(user
+        ? [
+            {
+              $lookup: {
+                from: "reaction",
+                let: { rateId: "$_id" },
+                pipeline: [
+                  {
+                    $match: {
+                      $expr: {
+                        $and: [
+                          { $eq: ["$rateId", "$$rateId"] },
+                          { $eq: ["$userId", user._id] },
+                        ],
+                      },
+                    },
+                  },
+                ],
+                as: "myReaction",
+              },
+            },
+            {
+              $addFields: {
+                myReaction: {
+                  $arrayElemAt: ["$myReaction.reaction", 0],
+                },
+              },
+            },
+          ]
+        : []),
+    ]).toArray();
```

Comment:
1. Refactor upper code block and try to avoid the log `? :` cause.
2. Currently we don't have reaction, please remove the aggregation with `reaction` table.